### PR TITLE
Add Context for LRO and paging and add related tests

### PR DESCRIPTION
**Pull request by haolingdong-msft**
Changes:
1. `ClientMethodTempalte`: Pass in `Context` for LRO and paging. 
2. `JavaSettings`: Set default polling strategy as `new DefaultPollingStrategy<>({httpPipeline}, null, {context})`
3. Add related tests for paging and LRO. The tests are used to make sure tracing `Context` is passed to policy correctly during multiple HTTP calls for LRO and paging.
4. Regen for protocol tests and vanilla tests.
5. Update `protocol-sdk-integration-tests/eng/versioning/verison_clients.txt` azure-core version to the latest.

### CODE REVIEW

File: protocol-tests/pom.xml [hunk 1]
```diff
@@ -72,6 +72,48 @@
...
+    <dependency>
+      <groupId>io.opentelemetry</groupId>
+      <artifactId>opentelemetry-exporter-logging</artifactId>
+      <version>1.11.0</version> <!-- {x-version-update;io.opentelemetry:opentelemetry-exporter-logging;external_dependency} -->
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>com.azure</groupId>
+      <artifactId>azure-security-keyvault-secrets</artifactId>
+      <version>4.3.8</version> <!-- {x-version-update;com.azure:azure-security-keyvault-secrets;dependency} -->
```
Comment (weidongxu-microsoft) [inline protocol-tests/pom.xml:84]: remove `<!-- {x-version-update;io.opentelemetry:opentelemetry-exporter-logging;external_dependency} -->`

and why keyvault?

File: protocol-tests/src/test/java/fixtures/dpgcustomization/DpgCustomizationTests.java [hunk 2]
```diff
@@ -213,4 +308,20 @@ public void testSendRequestMethod() {
...
+    static class TracingValidationPolicy implements HttpPipelinePolicy {
+
+        public HttpPipelineCallContext context;
+
+        TracingValidationPolicy() {
+        }
+
+        @Override
+        public Mono<HttpResponse> process(HttpPipelineCallContext httpPipelineCallContext, HttpPipelineNextPolicy httpPipelineNextPolicy) {
+            this.context = httpPipelineCallContext;
```
Comment (weidongxu-microsoft) [inline protocol-tests/src/test/java/fixtures/dpgcustomization/DpgCustomizationTests.java:321]: It should validate that every "context" has certain key, not just log the last "context".

File: protocol-tests/src/test/java/fixtures/dpgcustomization/DpgCustomizationTests.java [hunk 3]
```diff
@@ -94,6 +109,36 @@ public void getPages() {
...
+        Span userParentSpan = TRACER.spanBuilder("user-parent-span").startSpan();
+
+        Context context = new Context(
+                com.azure.core.util.tracing.Tracer.PARENT_TRACE_CONTEXT_KEY,
+                io.opentelemetry.context.Context.current().with(userParentSpan));
+
+        RequestOptions requestOptions = new RequestOptions();
+        requestOptions.setContext(context);
+
+        PagedFlux<BinaryData> pagedFlux = asyncClient.getPages("model", requestOptions);
```
Comment (weidongxu-microsoft) [inline protocol-tests/src/test/java/fixtures/dpgcustomization/DpgCustomizationTests.java:128]: Also test syncClient. They are different, hence focus is syncClient, not asyncClient.

File: javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java [hunk 4]
```diff
@@ -91,9 +92,9 @@ protected static void addOptionalAndConstantVariables(JavaBlock function, Client
      * @param clientMethod
      * @param proxyMethodAndConstantParameters
      * @param settings
-     * @param addOptional Whether add optional variables, init to default or null
-     * @param addConstant Whether add constant variables, init to default
-     * @param ignoreParameterNeedConvert When adding optional/constant variable, ignore those which need conversion from client type to wire type. Let "ConvertClientTypesToWireTypes" handle them.
+     * @param addOptional                      Whether add optional variables, init to default or null
+     * @param addConstant                      Whether add constant variables, init to default
+     * @param ignoreParameterNeedConvert       When adding optional/constant variable, ignore those which need conversion from client type to wire type. Let "ConvertClientTypesToWireTypes" handle them.
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java:97]: Please avoid all these unnecessary format change. It is very hard to find the actual code change.

File: javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java [hunk 5]
```diff
@@ -288,8 +295,8 @@ protected static void convertClientTypesToWireTypes(JavaBlock function, ClientMe
                             }
                         } else {
                             expression = String.format("Optional.ofNullable(%s).map(Collection::stream)" +
-                                    ".orElseGet(Stream::empty).map((item) -> Objects.toString(item, \"\"))" +
-                                    ".collect(Collectors.toList())",
+                                            ".orElseGet(Stream::empty).map((item) -> Objects.toString(item, \"\"))" +
+                                            ".collect(Collectors.toList())",
                                     parameterName);
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java:300]: revert it.

File: protocol-tests/pom.xml [hunk 6]
```diff
@@ -72,6 +72,18 @@
...
+      <artifactId>opentelemetry-exporter-logging</artifactId>
+      <version>1.0.0</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>com.azure</groupId>
+      <artifactId>azure-core-tracing-opentelemetry</artifactId>
+      <version>1.0.0-beta.21</version>
+      <scope>test</scope>
+    </dependency>
```
Comment (weidongxu-microsoft) [inline protocol-tests/pom.xml:86]: Are them still needed?

File: javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java [hunk 7]
```diff
@@ -595,9 +617,9 @@ protected void generateSyncMethod(ClientMethod clientMethod, JavaType typeBlock,
     /**
      * Generate javadoc for client method.
      *
-     * @param clientMethod client method
-     * @param typeBlock code block
-     * @param restAPIMethod proxy method
+     * @param clientMethod     client method
+     * @param typeBlock        code block
+     * @param restAPIMethod    proxy method
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java:622]: same, revert

File: javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java [hunk 8]
```diff
@@ -426,7 +434,11 @@ public final void write(ClientMethod clientMethod, JavaType typeBlock) {
                 break;
 
             case LongRunningBeginAsync:
-                generateLongRunningBeginAsync(clientMethod, typeBlock, restAPIMethod, settings);
+                if (settings.isLowLevelClient()) {
+                    generateProtocolLongRunningBeginAsync(clientMethod, typeBlock, restAPIMethod, settings);
+                } else {
+                    generateLongRunningBeginAsync(clientMethod, typeBlock, restAPIMethod, settings);
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java:440]: Please check if you need to handle the same issue for vanilla as well. Context need to be passed to the Strategy as well (maybe in `generateLongRunningBeginAsync`).

Comment (haolingdong-msft) [reply javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java:440]: ok, the test is used for validate context is passed in  correctly to the policy.

Comment (weidongxu-microsoft) [reply javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java:440]: I mean you do not need to add more test for vanilla, but make sure generated code is correct with `context` set to `Strategy` instances.

File: javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java [hunk 9]
```diff
@@ -803,6 +816,26 @@ protected void generateLongRunningBeginAsync(ClientMethod clientMethod, JavaType
...
+        typeBlock.annotation("ServiceMethod(returns = ReturnType.LONG_RUNNING_OPERATION)");
+        writeMethod(typeBlock, clientMethod.getMethodVisibility(), clientMethod.getDeclaration(), function -> {
+            function.line("return PollerFlux.create(Duration.ofSeconds(%s),", clientMethod.getMethodPollingDetails().getPollIntervalInSeconds());
+            function.increaseIndent();
+            function.line("() -> this.%s(%s),", clientMethod.getProxyMethod().getSimpleAsyncRestResponseMethodName(), clientMethod.getArgumentList());
+            function.line(pollingStrategy + ",");
+            TemplateUtil.writeLongRunningOperationTypeReference(function, clientMethod);
+            function.decreaseIndent();
+        });
+    }
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java:837]: Does it need to be similar to `generateLongRunningBeginAsync` method? As there is `polling` options that can allow dev to supply different `PollingStrategy` other than the `DefaultPollingStrategy`.

File: javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java [hunk 10]
```diff
@@ -426,15 +426,18 @@ public final void write(ClientMethod clientMethod, JavaType typeBlock) {
...
+                    generateLongRunningBeginAsync(clientMethod, typeBlock, restAPIMethod, settings);
+                }
                 break;
 
             case LongRunningBeginSync:
                 generateLongRunningBeginSync(clientMethod, typeBlock, restAPIMethod, settings);
                 break;
 
             case LongRunningSync:
-                generateSyncMethod(clientMethod, typeBlock, restAPIMethod, settings);
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java:437]: why removed?

File: javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java [hunk 11]
```diff
@@ -488,14 +491,24 @@ protected void generatePagingAsync(ClientMethod clientMethod, JavaType typeBlock
...
-                            clientMethod.getMethodPageDetails().getNextMethod().getProxyMethod().getPagingAsyncSinglePageMethodName(),
-                            clientMethod.getMethodPageDetails().getNextMethod().getArgumentListWithoutRequestOptions());
+                    if (settings.isLowLevelClient()) {
+                        function.line("nextLink -> %s(%s));",
+                                clientMethod.getMethodPageDetails().getNextMethod().getProxyMethod().getPagingAsyncSinglePageMethodName(),
+                                clientMethod.getMethodPageDetails().getNextMethod().getArgumentList().replace("requestOptions", "requestOptionsForNextPage"));
+                    } else {
+                        function.line("nextLink -> %s(%s));",
+                                clientMethod.getMethodPageDetails().getNextMethod().getProxyMethod().getPagingAsyncSinglePageMethodName(),
+                                clientMethod.getMethodPageDetails().getNextMethod().getArgumentListWithoutRequestOptions());
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java:510]: Is this same as `getArgumentList` if not LLC? And probably remove/comment-out `getArgumentListWithoutRequestOptions` method?

File: javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java [hunk 12]
```diff
@@ -803,6 +816,30 @@ protected void generateLongRunningBeginAsync(ClientMethod clientMethod, JavaType
...
+     * @param typeBlock type block
+     */
+    protected void generateProtocolLongRunningBeginAsync(ClientMethod clientMethod, JavaType typeBlock) {
+        String contextParam = "requestOptions != null ? requestOptions.getContext() : Context.NONE";
+        String pollingStrategy = clientMethod.getMethodPollingDetails().getPollingStrategy()
+                .replace("{httpPipeline}", clientMethod.getClientReference() + ".getHttpPipeline()")
+                .replace("{context}", contextParam)
+                .replace("{serializerAdapter}", clientMethod.getClientReference() + ".getSerializerAdapter()")
+                .replace("{intermediate-type}", clientMethod.getMethodPollingDetails().getIntermediateType().toString())
+                .replace("{final-type}", clientMethod.getMethodPollingDetails().getFinalType().toString());
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java:832]: nit, maybe share this code in a private method as duplicate in `generateLongRunningBeginAsync` method?

File: protocol-tests/src/main/java/fixtures/lro/implementation/LroRetrysImpl.java [hunk 13]
```diff
@@ -325,7 +325,10 @@
         return PollerFlux.create(
                 Duration.ofSeconds(1),
                 () -> this.put201CreatingSucceeded200WithResponseAsync(requestOptions),
-                new DefaultPollingStrategy<>(this.client.getHttpPipeline()),
+                new DefaultPollingStrategy<>(
+                        this.client.getHttpPipeline(),
+                        null,
+                        requestOptions != null ? requestOptions.getContext() : Context.NONE),
```
Comment (weidongxu-microsoft) [inline protocol-tests/src/main/java/fixtures/lro/implementation/LroRetrysImpl.java:331]: Should it be `requestOptions != null && requestOptions.getContext() != null ? requestOptions.getContext() : Context.NONE)`

I am not sure if `HttpClient.send(request, null)` works OK for all implementation of `HttpClient`.

File: javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java (hunk 11 above)
Comment (haolingdong-msft) [reply javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java:511]: should be the same for non-llc, I will comment-out `getArgumentListWithoutRequestOptions` and make llc, non-llc consistemt.

**Pull request merged**